### PR TITLE
Lazy evaluation of and, or, &, | 

### PR DESCRIPTION
**Pull request by smith120bh**
Fixes #2766

As discussed, this also adds `rawArgs` functionality to `OperatorNode` more generally - meaning that if any built-in function equivalent of any operator is overridden by a MathJS user, and the user sets the `rawArgs` property, that property will still be respected for the operator.

Tests and documentation added - but let me know if the structure / language should be updated, or if I missed somewhere!

### CODE REVIEW

File: test/unit-tests/expression/parse.test.js [hunk 1]
```diff
@@ -1465,6 +1465,10 @@ describe('parse', function () {
       assert.strictEqual(parseAndEval('true & false'), 0)
       assert.strictEqual(parseAndEval('false & true'), 0)
       assert.strictEqual(parseAndEval('false & false'), 0)
+
+      assert.strictEqual(parseAndEval('0 & undefined'), 0)
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:1469]: Using undefined here is a nifty way to verify that the second condition is not executed 😄, nice! Indeed we need to verify that `undefined` is not supported too and you did that with the `assert.throws(...)`.

**Review by josdejong — COMMENTED** [review]
Thanks a lot Brooks, this PR looks really solid. I made a couple of inline comments, can you have a look at those?

File: test/unit-tests/expression/parse.test.js [hunk 2]
```diff
@@ -1465,6 +1465,10 @@ describe('parse', function () {
...
+      assert.strictEqual(parseAndEval('false & undefined'), 0)
+      assert.throws(function () { parseAndEval('true & undefined') }, TypeError)
```
Comment (josdejong) [inline test/unit-tests/expression/parse.test.js:1471]: Can you also add unit tests that validate defining variables, like `b =2 and c = 3`, to verify whether the variables are put in the original scope correctly? I think the `createSubScope(scope, args)` in `OperatorNode` should not be there, but I'm not entirely sure.

File: src/expression/transform/and.transform.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+const name = 'and'
+const dependencies = ['typed', 'matrix', 'zeros', 'add', 'equalScalar', 'not', 'concat']
+
+export const createAndTransform = /* #__PURE__ */ factory(name, dependencies, ({ typed, matrix, equalScalar, zeros, not, concat }) => {
+  const and = createAnd({ typed, matrix, equalScalar, zeros, not, concat })
+
+  function andTransform (args, math, scope) {
+    const condition1 = args[0].compile().evaluate(scope)
+    if (isNumber(condition1) || isBoolean(condition1) || isBigNumber(condition1) || isComplex(condition1) || isUnit(condition1)) {
```
[the remark this answers is not in the document]
Comment (smith120bh) [reply src/expression/transform/and.transform.js:14]: That is indeed a clever way to replace `testCondition` - however, it doesn't fully get rid of the type checks. It errors out if `condition1` is an array or matrix. However, if we just add an `if (!isCollection(condition1))`, then it'd work!

File: src/factoriesAny.js [hunk 4]
```diff
@@ -359,3 +359,7 @@ export { createQuantileSeqTransform } from './expression/transform/quantileSeq.t
 export { createCumSumTransform } from './expression/transform/cumsum.transform.js'
 export { createVarianceTransform } from './expression/transform/variance.transform.js'
 export { createPrintTransform } from './expression/transform/print.transform.js'
+export { createAndTransform } from './expression/transform/and.transform.js'
```
[the remark this answers is not in the document]
Comment (smith120bh) [reply src/factoriesAny.js:362]: Unfortunately not - it's not logically possible to lazy evaluate an `xor`; you need to know both arguments to determine anything about an `xor`'s results.

File: src/function/bitwise/bitAnd.js [hunk 5]
```diff
@@ -23,6 +23,7 @@ export const createBitAnd = /* #__PURE__ */ factory(name, dependencies, ({ typed
   /**
    * Bitwise AND two values, `x & y`.
    * For matrices, the function is evaluated element wise.
+   * For non-matrices, the function is evaluated lazily.
```
[the remark this answers is not in the document]
Comment (smith120bh) [reply src/function/bitwise/bitAnd.js:26]: Makes sense. I thought having a whole new sections felt excessive, so I've only added a couple sentences to the "Precedence" section - but I have no strong feelings here. At the end of the day, lazy evaluation still evaluates to the same answer, so it doesn't feel like it deserves too much space?

File: src/expression/transform/and.transform.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (josdejong) [reply src/expression/transform/and.transform.js]: Good point about handling collections. Your solution checking `isCollection` makes sense indeed 👍 .

File: test/unit-tests/expression/parse.test.js (hunk 2 above)
Comment (josdejong) [reply test/unit-tests/expression/parse.test.js:1471]: @smith120bh can you still have a look into adding tests with defining variables? Besides that I think the PR is ready to be merged, it looks really neat 👌 .

Comment (josdejong) [reply test/unit-tests/expression/parse.test.js:1471]: ping @smith120bh any chance of looking into some tests with defining variables? (If not I can have a look)

Comment (josdejong) [reply test/unit-tests/expression/parse.test.js:1471]: Addressed via e4f8f78809f6f09046974279bf0154a041fb7d72

**Pull request closed without merging**